File: server_py/abi/providers/llm/gemini.py
"""Google Gemini akisli uretim (streamGenerateContent, SSE)."""

from __future__ import annotations

import json
import logging
import re
from typing import Any, AsyncIterator

import httpx

from ...config import config
from .base import LlmRequest

log = logging.getLogger("abi.gemini")

_MODEL_ISSUE = re.compile(r"model|not found|unsupported", re.IGNORECASE)


class GeminiError(Exception):
    def __init__(self, message: str, status: int | None = None, streamed: bool = False) -> None:
        super().__init__(message)
        self.status = status
        self.streamed = streamed


def _contents(request: LlmRequest) -> list[dict[str, Any]]:
    contents: list[dict[str, Any]] = [
        {"role": "user" if turn.role == "user" else "model", "parts": [{"text": turn.text}]}
        for turn in request.history
    ]

    # Goruntuler son kullanici mesajina eklenir: model once bakar, sonra okur.
    parts: list[dict[str, Any]] = [
        {"inlineData": {"mimeType": frame.mime, "data": frame.data}} for frame in request.frames
    ]
    parts.append({"text": request.message})
    contents.append({"role": "user", "parts": parts})
    return contents


def _body(request: LlmRequest, *, thinking: bool) -> dict[str, Any]:
    generation: dict[str, Any] = {
        "temperature": request.temperature,
        "maxOutputTokens": request.max_tokens,
        "topP": 0.95,
    }
    if not thinking:
        # Karakter kisa konusuyor; dusunme butcesi gecikmeyi katliyor.
        generation["thinkingConfig"] = {"thinkingBudget": 0}

    return {
        "systemInstruction": {"parts": [{"text": request.system}]},
        "contents": _contents(request),
        "generationConfig": generation,
        "safetySettings": [
            {"category": category, "threshold": "BLOCK_ONLY_HIGH"}
            for category in (
                "HARM_CATEGORY_HARASSMENT",
                "HARM_CATEGORY_HATE_SPEECH",
                "HARM_CATEGORY_SEXUALLY_EXPLICIT",
                "HARM_CATEGORY_DANGEROUS_CONTENT",
            )
        ],
    }


class GeminiProvider:
    id = "gemini"
    label = "Google Gemini"
    vision = True

    def __init__(self) -> None:
        # Calistigi dogrulanan model; sonraki isteklerde dogrudan kullanilir.
        self._resolved: str | None = None

    async def available(self) -> bool:
        return bool(config.gemini.api_key)

    def _model_chain(self) -> list[str]:
        if self._resolved:
            return [self._resolved]
        chain = [config.gemini.model, *config.gemini.fallback_models]
        seen: list[str] = []
        for model in chain:
            if model and model not in seen:
                seen.append(model)
        return seen

    async def stream(self, request: LlmRequest) -> AsyncIterator[str]:
        if not config.gemini.api_key:
            raise GeminiError("GEMINI_API_KEY tanimli degil")

        last: GeminiError | None = None
        for model in self._model_chain():
            try:
                produced = False
                async for piece in self._stream_model(model, request, thinking=False):
                    produced = True
                    yield piece
                self._resolved = model
                if not produced:
                    log.warning("%s bos cevap dondurdu", model)
                return
            except GeminiError as err:
                # Akis basladiktan sonra model degistirmek cevabi bozar.
                if err.streamed:
                    raise
                last = err
                retryable = err.status in (404, 429) or (
                    err.status == 400 and bool(_MODEL_ISSUE.search(str(err)))
                )
                log.warning("%s basarisiz (%s) %s", model, err.status, str(err)[:200])
                if not retryable:
                    break
        raise last or GeminiError("model bulunamadi")

    async def _stream_model(
        self, model: str, request: LlmRequest, *, thinking: bool
    ) -> AsyncIterator[str]:
        url = f"{config.gemini.base_url}/models/{model}:streamGenerateContent?alt=sse"
        headers = {"content-type": "application/json", "x-goog-api-key": config.gemini.api_key}

        async with httpx.AsyncClient(timeout=httpx.Timeout(60.0, connect=10.0)) as client:
            async with client.stream(
                "POST", url, headers=headers, json=_body(request, thinking=thinking)
            ) as response:
                if response.status_code >= 400:
                    detail = (await response.aread()).decode("utf-8", "replace")
                    # thinkingConfig eski modellerde reddediliyor; sade govdeyle tekrar dene.
                    if response.status_code == 400 and "thinking" in detail.lower() and not thinking:
                        async for piece in self._stream_model(model, request, thinking=True):
                            yield piece
                        return
                    raise GeminiError(detail[:300] or response.reason_phrase, response.status_code)

                started = False
                try:
                    async for line in response.aiter_lines():
                        if not line.startswith("data:"):
                            continue
                        payload = line[5:].strip()
                        if not payload or payload == "[DONE]":
                            continue
                        try:
                            chunk = json.loads(payload)
                        except json.JSONDecodeError:
                            continue

                        feedback = chunk.get("promptFeedback") or {}
                        if feedback.get("blockReason"):
                            log.warning("istek engellendi (%s)", feedback["blockReason"])

                        candidates = chunk.get("candidates") or []
                        if not candidates:
                            continue
                        for part in (candidates[0].get("content") or {}).get("parts") or []:
                            text = part.get("text")
                            if text:
                                started = True
                                yield text
                except httpx.HTTPError as err:
                    raise GeminiError(str(err), streamed=started) from err

File: server/src/providers/llm/gemini.ts
import { config } from '../../config.js';
import type { LlmProvider, LlmRequest } from '../types.js';
import { log } from '../../util/log.js';

interface GeminiPart {
  text?: string;
}
interface GeminiCandidate {
  content?: { parts?: GeminiPart[] };
  finishReason?: string;
}
interface GeminiStreamChunk {
  candidates?: GeminiCandidate[];
  promptFeedback?: { blockReason?: string };
}

/** Denenecek model sirasi: env tercihi once, sonra bilinen fallback zinciri. */
function modelChain(): string[] {
  const chain = [config.gemini.model, ...config.gemini.fallbackModels];
  return [...new Set(chain.filter(Boolean))];
}

function toContents(req: LlmRequest) {
  const contents = req.history.map((turn) => ({
    role: turn.role === 'user' ? 'user' : 'model',
    parts: [{ text: turn.text }],
  }));
  contents.push({ role: 'user', parts: [{ text: req.message }] });
  return contents;
}

/**
 * Gemini streamGenerateContent (SSE). Karakter gecikmesi kritik oldugu icin
 * token'lar geldigi anda yukari akitilir.
 */
export class GeminiProvider implements LlmProvider {
  readonly id = 'gemini';
  readonly label = 'Google Gemini';

  /** Calistigi dogrulanan model; sonraki isteklerde dogrudan kullanilir. */
  private resolvedModel: string | null = null;

  async available(): Promise<boolean> {
    return Boolean(config.gemini.apiKey);
  }

  async *stream(req: LlmRequest): AsyncIterable<string> {
    if (!config.gemini.apiKey) throw new Error('GEMINI_API_KEY tanimli degil');

    const models = this.resolvedModel ? [this.resolvedModel] : modelChain();
    let lastError: Error | null = null;

    for (const model of models) {
      try {
        let produced = false;
        for await (const piece of this.streamModel(model, req)) {
          produced = true;
          yield piece;
        }
        this.resolvedModel = model;
        if (!produced) log.warn(`gemini: ${model} bos cevap dondurdu`);
        return;
      } catch (err) {
        const error = err as Error & { status?: number };
        // Akis basladiktan sonraki hatada model degistirmek cevabi bozar.
        if ((error as { streamed?: boolean }).streamed) throw error;
        if (error.name === 'AbortError') throw error;
        lastError = error;
        // Gecersiz anahtar gibi kalici hatalarda model degistirmek anlamsiz;
        // sadece "model yok / kota" durumlarinda zincire devam et.
        const modelIssue = /model|not found|unsupported/i.test(error.message);
        const retryable =
          error.status === 404 || error.status === 429 || (error.status === 400 && modelIssue);
        log.warn(`gemini: ${model} basarisiz (${error.status ?? '-'}) ${error.message}`);
        if (!retryable) break;
      }
    }
    throw lastError ?? new Error('gemini: model bulunamadi');
  }

  private async *streamModel(model: string, req: LlmRequest): AsyncIterable<string> {
    const url =
      `${config.gemini.baseUrl}/models/${encodeURIComponent(model)}:streamGenerateContent?alt=sse`;

    const res = await fetch(url, {
      method: 'POST',
      headers: {
        'content-type': 'application/json',
        'x-goog-api-key': config.gemini.apiKey,
      },
      signal: req.signal,
      body: JSON.stringify({
        systemInstruction: { parts: [{ text: req.system }] },
        contents: toContents(req),
        generationConfig: {
          temperature: req.temperature,
          maxOutputTokens: req.maxTokens,
          topP: 0.95,
          // Karakter kisa konusuyor; dusunme butcesi gecikmeyi katlar.
          thinkingConfig: { thinkingBudget: 0 },
        },
        safetySettings: [
          'HARM_CATEGORY_HARASSMENT',
          'HARM_CATEGORY_HATE_SPEECH',
          'HARM_CATEGORY_SEXUALLY_EXPLICIT',
          'HARM_CATEGORY_DANGEROUS_CONTENT',
        ].map((category) => ({ category, threshold: 'BLOCK_ONLY_HIGH' })),
      }),
    });

    if (!res.ok || !res.body) {
      const detail = await res.text().catch(() => '');
      const err = Object.assign(new Error(detail.slice(0, 300) || res.statusText), {
        status: res.status,
      });
      // thinkingConfig eski modellerde reddedilir; bir kez daha sade govde ile dene.
      if (res.status === 400 && detail.toLowerCase().includes('thinking')) {
        yield* this.streamModelPlain(model, req);
        return;
      }
      throw err;
    }

    yield* readSse(res.body, model);
  }

  /** thinkingConfig desteklemeyen modeller icin sade govde. */
  private async *streamModelPlain(model: string, req: LlmRequest): AsyncIterable<string> {
    const url =
      `${config.gemini.baseUrl}/models/${encodeURIComponent(model)}:streamGenerateContent?alt=sse`;
    const res = await fetch(url, {
      method: 'POST',
      headers: { 'content-type': 'application/json', 'x-goog-api-key': config.gemini.apiKey },
      signal: req.signal,
      body: JSON.stringify({
        systemInstruction: { parts: [{ text: req.system }] },
        contents: toContents(req),
        generationConfig: {
          temperature: req.temperature,
          maxOutputTokens: req.maxTokens,
          topP: 0.95,
        },
      }),
    });
    if (!res.ok || !res.body) {
      const detail = await res.text().catch(() => '');
      throw Object.assign(new Error(detail.slice(0, 300) || res.statusText), { status: res.status });
    }
    yield* readSse(res.body, model);
  }
}

async function* readSse(body: ReadableStream<Uint8Array>, model: string): AsyncIterable<string> {
  const reader = body.getReader();
  const decoder = new TextDecoder();
  let buffer = '';
  let started = false;

  try {
    while (true) {
      const { done, value } = await reader.read();
      if (done) break;
      buffer += decoder.decode(value, { stream: true });

      let nl: number;
      while ((nl = buffer.indexOf('\n')) !== -1) {
        const line = buffer.slice(0, nl).trim();
        buffer = buffer.slice(nl + 1);
        if (!line.startsWith('data:')) continue;
        const payload = line.slice(5).trim();
        if (!payload || payload === '[DONE]') continue;

        let parsed: GeminiStreamChunk;
        try {
          parsed = JSON.parse(payload) as GeminiStreamChunk;
        } catch {
          continue;
        }
        if (parsed.promptFeedback?.blockReason) {
          log.warn(`gemini: istek engellendi (${parsed.promptFeedback.blockReason})`);
        }
        const parts = parsed.candidates?.[0]?.content?.parts ?? [];
        for (const part of parts) {
          if (!part.text) continue;
          started = true;
          yield part.text;
        }
      }
    }
  } catch (err) {
    const error = err as Error;
    if (started) Object.assign(error, { streamed: true });
    throw error;
  } finally {
    reader.releaseLock();
    void model;
  }
}
